File: 166.fraction-to-recurring-decimal.py
#
# @lc app=leetcode id=166 lang=python3
#
# [166] Fraction to Recurring Decimal
#

# Time:  O(logn)
# Space: O(1)

# @lc code=start
class Solution:
    def fractionToDecimal(self, numerator: int, denominator: int) -> str:
        res = ""
        if numerator > 0 and denominator < 0 or numerator < 0 and denominator > 0:
            res = "-"

        abs_n, abs_d = abs(numerator), abs(denominator)
        res += str(abs_n // abs_d)
        abs_n %= abs_d

        if abs_n > 0:
            res += "."

        lookup = {}
        while abs_n and abs_n not in lookup:
            lookup[abs_n] = len(res)
            abs_n *= 10
            res += str(abs_n // abs_d)
            abs_n %= abs_d

        if abs_n in lookup:
            res = res[: lookup[abs_n]] + "(" + res[lookup[abs_n] :] + ")"
        return res


# @lc code=end

File: 166.fraction-to-recurring-decimal.js
/*
 * @lc app=leetcode id=166 lang=javascript
 *
 * [166] Fraction to Recurring Decimal
 */

// @lc code=start
/**
 * @param {number} numerator
 * @param {number} denominator
 * @return {string}
 */
var fractionToDecimal = function (numerator, denominator) {
  let res =
    (numerator > 0 && denominator < 0) || (numerator < 0 && denominator > 0)
      ? "-"
      : "";
  let absN = Math.abs(numerator);
  let absD = Math.abs(denominator);
  res += `${Math.trunc(absN / absD)}`;
  absN = (absN % absD) * 10;
  if (absN > 0) {
    res += ".";
  }

  const lookup = {};
  while (absN && !(absN in lookup)) {
    lookup[absN] = res.length;
    res += `${Math.trunc(absN / absD)}`;
    absN = (absN % absD) * 10;
  }
  if (absN in lookup) {
    return `${res.slice(0, lookup[absN])}(${res.slice(lookup[absN])})`;
  }
  return res;
};
// @lc code=end
